Userland/Demos/VirGLDemo/VirGLDemo.cpp

```diff
@@ -0,0 +1,322 @@
+    float g;
+    float b;
+    float x;
+    float y;
+    float z;
+};
+
+int gpu_fd;
+ResourceID vbo_resource_id;
+ResourceID drawtarget;
+ResourceID depthbuffer_surface;
+ObjectHandle blend_handle;
+ObjectHandle drawtarget_surface_handle;
+ObjectHandle depthbuffer_surface_handle;
+ObjectHandle ve_handle;
+ObjectHandle frag_shader_handle;
+ObjectHandle vert_shader_handle;
+ObjectHandle rasterizer_handle;
+ObjectHandle dsa_handle;
+Vector<VertexData> g_vertices;
+
+static ObjectHandle allocate_handle()
+{
+    static u32 last_allocated_handle = 0;
+    return { ++last_allocated_handle };
+}
+
+static void upload_command_buffer(Vector<u32> const& command_buffer)
+{
+    VERIFY(command_buffer.size() <= AK::NumericLimits<u32>::max());
+    VirGLCommandBuffer command_buffer_descriptor {
+        .data = command_buffer.data(),
+        .num_elems = (u32)command_buffer.size(),
+    };
+    VERIFY(ioctl(gpu_fd, VIRGL_IOCTL_SUBMIT_CMD, &command_buffer_descriptor) >= 0);
+}
+
+static ResourceID create_virgl_resource(VirGL3DResourceSpec& spec)
+{
+    VERIFY(ioctl(gpu_fd, VIRGL_IOCTL_CREATE_RESOURCE, &spec) >= 0);
+    return spec.created_resource_id;
+}
+
+static Vector<VertexData> gen_vertex_data()
+{
+    Vector<VertexData> data;
+    static const VertexData vertices[8] = {
+        VertexData { .r = 0, .g = 0, .b = 0, .x = -0.5, .y = -0.5, .z = -0.5 },
+        VertexData { .r = 0, .g = 0, .b = 0, .x = 0.5, .y = -0.5, .z = -0.5 },
+        VertexData { .r = 0, .g = 0, .b = 0, .x = -0.5, .y = 0.5, .z = -0.5 },
+        VertexData { .r = 0, .g = 0, .b = 0, .x = 0.5, .y = 0.5, .z = -0.5 },
+        VertexData { .r = 0, .g = 0, .b = 0, .x = -0.5, .y = -0.5, .z = 0.5 },
+        VertexData { .r = 0, .g = 0, .b = 0, .x = 0.5, .y = -0.5, .z = 0.5 },
+        VertexData { .r = 0, .g = 0, .b = 0, .x = -0.5, .y = 0.5, .z = 0.5 },
+        VertexData { .r = 0, .g = 0, .b = 0, .x = 0.5, .y = 0.5, .z = 0.5 },
+    };
+    size_t tris[36] = {
+        0, 1, 2, 1, 3, 2, // Top
+        4, 0, 6, 0, 2, 6, // Left
+        4, 5, 0, 5, 1, 0, // Up
+        1, 5, 3, 5, 7, 3, // Right
+        2, 3, 6, 3, 7, 6, // Down
+        5, 4, 7, 4, 6, 7, // Bottom
+    };
+    for (auto index : tris) {
+        data.append(vertices[index]);
+    }
+    // Choose random colors for each face of the cube
+    for (auto i = 0; i < 6; ++i) {
+        float red = (rand() % 256) / 255.f;
+        float green = (rand() % 256) / 255.f;
+        float blue = (rand() % 256) / 255.f;
+        for (auto j = 0; j < 6; ++j) {
+            auto& vertex = data[i * 6 + j];
+            vertex.r = red;
+            vertex.g = green;
+            vertex.b = blue;
+        }
+    }
+    return data;
+}
+
+static void init()
+{
+    // Open the device
+    gpu_fd = open("/dev/gpu0", O_RDWR);
+    VERIFY(gpu_fd >= 0);
+    // Create a VertexElements resource
+    VirGL3DResourceSpec vbo_spec {
+        .target = AK::to_underlying(Gallium::PipeTextureTarget::BUFFER), // pipe_texture_target
```

Comment:
No need to qualify this with `AK::` (here and multiple other places).

Reply:
Fixed in next revision.